src/hotspot/share/compiler/directivesParser.cpp

```diff
@@ -320,29 +320,37 @@ bool DirectivesParser::set_option_flag(JSON_TYPE t, JSON_VAL* v, const key* opti
         s[v->str.length] = '\0';
         (set->*test)((void *)&s);
 
+        bool valid = true;
+
         if (strncmp(option_key->name, "ControlIntrinsic", 16) == 0) {
           ControlIntrinsicValidator validator(s, false/*disabled_all*/);
 
-          if (!validator.is_valid()) {
+          valid = validator.is_valid();
+          if (!valid) {
             error(VALUE_ERROR, "Unrecognized intrinsic detected in ControlIntrinsic: %s", validator.what());
-            return false;
           }
         } else if (strncmp(option_key->name, "DisableIntrinsic", 16) == 0) {
           ControlIntrinsicValidator validator(s, true/*disabled_all*/);
 
-          if (!validator.is_valid()) {
+          valid = validator.is_valid();
+          if (!valid) {
             error(VALUE_ERROR, "Unrecognized intrinsic detected in DisableIntrinsic: %s", validator.what());
-            return false;
           }
         } else if (strncmp(option_key->name, "PrintIdealPhase", 15) == 0) {
           uint64_t mask = 0;
           PhaseNameValidator validator(s, mask);
 
-          if (!validator.is_valid()) {
+          valid = validator.is_valid();
+          if (!valid) {
             error(VALUE_ERROR, "Unrecognized phase name detected in PrintIdealPhase: %s", validator.what());
-            return false;
+          } else {
+            set->set_ideal_phase_mask(mask);
           }
-          set->set_ideal_phase_mask(mask);
+        }
+
+        FREE_C_HEAP_ARRAY(char, s);
```

Comment:
This looks unsafe.  We shouldn't free the memory without clearing all references to it, otherwise there is a dangling pointer.  There is already another reference to the memory because of this call:

`(set->*test)((void *)&s);` (see the set_function_definition macro)

I think it would be better to move this copying call until after validation has been done.

Reply:
I am very confused actually, `(set->*test)((void *)&s);` calls DirectiveSet::set_X. Looking at the code, it simply just stores the pointer? Is the DirectiveSet supposed to own the option? And if so, who is freeing it then? It doesn't look like cloning actually clones the underlying storage and the DirectiveSet destructor doesn't free it.

So really DirectiveSet::~DirectiveSet should be freeing the string storage and DirectiveSet::set_X is taking ownership. Yeah?

Reply:
Right, the ownership of the string seems unclear.  If we try to free the string storage in ~DirectiveSet I think it will result in a double free, because clone() appears to do a shallow copy.

Reply:
Worse yet, it looks like the default value can point to the existing storage of the DisableIntrinsic and ControlIntrinsic global values, making it doubly hard to know if the storage can be freed, unless we get rid of simple = assignment and always allocate new memory when assigning the string values.
But why re-invent the wheel?  Can't we use something like C++ std::string to manage the lifetimes for us?

Reply:
Or maybe ~DirectiveSet could use the _modified array to decide if it should free the string flags.

Reply:
`DirectiveSet::compilecommand_compatibility_init` makes things even more complicated, because it doesnt update `_modified` and it should probably be making a copy of the string provided via `CompilerOracle::has_option_value`. Ugh...

Reply:
Nevermind, it looks like `CompilerOracle::has_option_value` has static duration effectively. So I should be able to fix this.

Reply:
Okay, think I was able to figure it out. Running GHA again.